src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -4485,11 +4514,17 @@ FindReferencedTableColumn(Expr *columnExpression, List *parentQueryList, Query *
 		Expr *joinColumn = list_nth(joinColumnList, joinColumnIndex);
 
 		/* parent query list stays the same since still in the same query boundary */
-		FindReferencedTableColumn(joinColumn, parentQueryList, query,
-								  relationId, column);
+		FindReferencedTableColumn(joinColumn, parentQueryList, query, column,
+								  rteContainingReferencedColumn, skipOuterVars);
 	}
 	else if (rangeTableEntry->rtekind == RTE_CTE)
 	{
+		/*
```

Comment:
@SaitTalhaNisanci this hits an assertion failure on multi-mx-schedule on coordinator_should_have_shards mx test:

```SQL
WITH a AS (SELECT * FROM table_1 ORDER BY 1,2 DESC LIMIT 1)
 INSERT INTO table_1 SELECT count(*),
 key
 FROM a JOIN table_2 USING (key)
 GROUP BY key
 HAVING (max(table_2.value) >= (SELECT value FROM a));
```

Reply:
@onderkalaci looking at this assert, I am curious why we have added it. Because if `skipOuterVars` is true then we don't modify the `parentQueryList`:
![image](https://user-images.githubusercontent.com/17179875/129165480-f663e4f7-ed1d-4e2e-85ad-15e2c3d4d797.png)

And we haven't changed the CTE block, so when `skipOuterVars` is true, the function will be executed same as the previous version. In that case it seems like we remove this assert?

Reply:
I think when `! skipOuterVars ` we modify `parentQueryList` and `parentQueryList` is used when `RTE_CTE`.

I think one reason we added the assert was we weren't sure if  `rangeTableEntry->ctelevelsup > 0` and `candidateColumn->varlevelsup > 0` can happen at the same time.

If that happens, our logic could be broken for `RTE_CTE`.

@SaitTalhaNisanci does that make sense?